### ISSUE TITLE
Local versus global Redux

### ISSUE THREAD

**Issue by lastmjs** [lastmjs/javascript-practice#45]
I think I have a path forward for easy local component updates through Redux, without having to create a component id. We'll compose reducers on the store. Each component will also have its own reducer that supports a LOCAL_UPDATE action. The type is LOCAL_UPDATE, the key is the local key, and the value the value. The global state will have two top level properties, global and local. I'm not sure if this maintains the replayability that we might want to retain...but hopefully it does. The problem is that each component will not be accounted for on the global state. Each component will essentially have its own state composed with the global state...not ideal

**Comment by bmackley** [comment]
So as I've learned a bit of React, I'm still miffed at the idea of why we need redux to manage the state of each individual component? Shouldn't a component be able to handle it's own state? Ideally a component is going to be open source then it will be a dumb component that only accepts state through properties and only changes state through events? Is that how it works with the question elements?

**Comment by lastmjs** [comment]
No, the question elements have a lot of complex state interactions, that's why they have their own redux store. If a component gets complex enough, its state management will start to get unreasonable. I've come to the conclusion that the only way is to fully manage side effects within classes, right now through Redux. There should never be a mutation in a class, and I'm working very hard to make it impossible in our code base

**Comment by lastmjs** [comment]
You're going to have to trust me on this. Our duty is software developers is to manage complexity. State interactions are some of the largest contributors to complexity. It starts to get out of control very quickly. It must be managed, no mutations in classes, it's evil to do so

**Comment by lastmjs** [comment]
The biggest thing here is, like you pointed out, components have their own state. If their own state gets complex enough, it needs something like Redux or you will die. That's why I'm working to find a way nearly as easy as straight up mutations within the class, but tracked by Redux and updated only through the render method.

**Comment by lastmjs** [comment]
Create a module that can be imported into each component. This module exports a proxy, which creates a unique identifier and will intercept all sets. Each set will take the property, value, and component id and dispatch a local state update against the global Redux store. We can have a global and local state property on the Store...perhaps we can somehow merge the local state into the global state for each component. We will then set properties on each component's state object. The most important thing is maintaining unidirectional data flow and tracking of mutations. I believe a paradigm similar to this will lead to an ideal situation
